File: omrchecker-js/packages/core/src/processors/threshold/GlobalThreshold.ts
/**
 * Global threshold strategy for bubble detection.
 *
 * TypeScript port of src/processors/threshold/GlobalThreshold.py
 * Maintains 1:1 correspondence with Python implementation.
 */

import { Logger } from '../../utils/logger';

const logger = new Logger('GlobalThreshold');

/**
 * Configuration for threshold calculation
 */
export interface ThresholdConfig {
  defaultThreshold: number;
  minJump: number;
  jumpDelta?: number;
  minGapTwoBubbles?: number;
  minJumpSurplusForGlobalFallback?: number;
  confidentJumpSurplusForDisparity?: number;
  globalThresholdMargin?: number;
  outlierDeviationThreshold?: number;
}

/**
 * Result of threshold calculation
 */
export interface ThresholdResult {
  thresholdValue: number;
  confidence: number;
  maxJump: number;
  methodUsed: string;
  fallbackUsed: boolean;
  metadata: Record<string, any>;
}

/**
 * Global threshold strategy - finds the first large gap in sorted bubble values.
 *
 * This is the primary threshold detection method used in OMRChecker.
 */
export class GlobalThreshold {
  calculateThreshold(
    bubbleMeanValues: number[],
    config: ThresholdConfig
  ): ThresholdResult {
    if (bubbleMeanValues.length < 2) {
      return {
        thresholdValue: config.defaultThreshold,
        confidence: 0.0,
        maxJump: 0.0,
        methodUsed: 'global_default',
        fallbackUsed: true,
        metadata: {},
      };
    }

    const sortedValues = [...bubbleMeanValues].sort((a, b) => a - b);

    // Find the FIRST LARGE GAP using looseness parameter
    const looseness = 1;
    const ls = Math.floor((looseness + 1) / 2);

    let maxJump = 0;
    let threshold = config.defaultThreshold;
    let bestIndex = -1;

    for (let i = ls; i < sortedValues.length - ls; i++) {
      const jump = sortedValues[i] - sortedValues[i - 1];
      if (jump > maxJump) {
        maxJump = jump;
        bestIndex = i;
      }
    }

    if (bestIndex !== -1) {
      threshold = (sortedValues[bestIndex - 1] + sortedValues[bestIndex]) / 2;
    }

    // Calculate confidence based on jump size
    let confidence = 0.0;
    if (maxJump > config.minJump) {
      confidence = Math.min(1.0, maxJump / (config.minJump * 2));
    }

    logger.debug(
      `GlobalThreshold: maxJump=${maxJump.toFixed(2)}, threshold=${threshold.toFixed(2)}, confidence=${confidence.toFixed(2)}`
    );

    return {
      thresholdValue: threshold,
      confidence,
      maxJump,
      methodUsed: 'global_threshold',
      fallbackUsed: maxJump <= config.minJump,
      metadata: {
        sortedValues,
        bestIndex,
      },
    };
  }
}

File: src/processors/threshold/global_threshold.py
"""Global threshold strategy for bubble detection.

Strategy using global file-level statistics.
Based on the existing get_global_threshold logic.
Finds the largest gap in sorted bubble means across all fields.
"""

from src.processors.threshold.threshold_result import ThresholdConfig, ThresholdResult
from src.processors.threshold.threshold_strategy import ThresholdStrategy


class GlobalThresholdStrategy(ThresholdStrategy):
    """Strategy using global file-level statistics.

    Based on the existing get_global_threshold logic.
    Finds the largest gap in sorted bubble means across all fields.
    """

    def calculate_threshold(
        self, bubble_mean_values: list[float], config: ThresholdConfig
    ) -> ThresholdResult:
        """Calculate global threshold by finding largest gap."""
        if len(bubble_mean_values) < 2:
            return ThresholdResult(
                threshold_value=config.default_threshold,
                confidence=0.0,
                max_jump=0.0,
                method_used="global_default",
                fallback_used=True,
            )

        sorted_values = sorted(bubble_mean_values)

        # Find the FIRST LARGE GAP using looseness
        looseness = 1
        ls = (looseness + 1) // 2
        total_bubbles_loose = len(sorted_values) - ls

        max_jump = config.min_jump
        threshold = config.default_threshold

        for i in range(ls, total_bubbles_loose):
            jump = sorted_values[i + ls] - sorted_values[i - ls]
            if jump > max_jump:
                max_jump = jump
                threshold = sorted_values[i - ls] + jump / 2

        # Calculate confidence based on jump size
        # Higher jump = higher confidence
        confidence = min(1.0, max_jump / (config.min_jump * 3))

        return ThresholdResult(
            threshold_value=threshold,
            confidence=confidence,
            max_jump=max_jump,
            method_used="global_max_jump",
            fallback_used=max_jump < config.min_jump,
            metadata={
                "num_bubbles": len(bubble_mean_values),
                "min_value": min(bubble_mean_values),
                "max_value": max(bubble_mean_values),
            },
        )
